larpandoracontent/LArControlFlow/NeutrinoIdTool.cc

```diff
@@ -10,30 +10,464 @@
+
+    if (nuNHitsTotal > 0)
+        completeness = static_cast<float>(nNuHitsInBestSlice) / static_cast<float>(nuNHitsTotal);
+
+    return this->PassesQualityCuts(pAlgorithm, purity, completeness);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool NeutrinoIdTool::PassesQualityCuts(const MasterAlgorithm *const pAlgorithm, const float purity, const float completeness) const
+{
+    if (purity < m_minPurity || completeness < m_minCompleteness) return false;
+    if (m_selectNuanceCode && (this->GetNuanceCode(pAlgorithm) != m_nuance)) return false;
+
+    return true;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void NeutrinoIdTool::Collect2DHits(const PfoList &pfos, CaloHitList &hitList) const
+{
+    CaloHitList collectedHits;
+    LArPfoHelper::GetCaloHits(pfos, TPC_VIEW_U, collectedHits);
+    LArPfoHelper::GetCaloHits(pfos, TPC_VIEW_V, collectedHits);
+    LArPfoHelper::GetCaloHits(pfos, TPC_VIEW_W, collectedHits);
+
+    for (const CaloHit *const pCaloHit : collectedHits)
+    {
+        // ATTN hits collected from Pfos are copies of hits passed from master instance, we need to access their parent to use MC info
+        const CaloHit *pParentCaloHit(static_cast<const CaloHit *>(pCaloHit->GetParentAddress()));
+
+        // Ensure no hits have been double counted
+        if (std::find(hitList.begin(), hitList.end(), pParentCaloHit) == hitList.end())
+            hitList.push_back(pParentCaloHit);
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+unsigned int NeutrinoIdTool::CountNeutrinoInducedHits(const CaloHitList &hitList) const
+{
+    unsigned int nNuHits(0);
+    for (const CaloHit *const pCaloHit : hitList)
+    {
+        try
+        {
+            if (LArMCParticleHelper::IsNeutrino(LArMCParticleHelper::GetParentMCParticle(MCParticleHelper::GetMainMCParticle(pCaloHit))))
+                nNuHits++;
+        }
+        catch (StatusCodeException &)
+        {
+        }
+    }
+
+    return nNuHits;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+int NeutrinoIdTool::GetNuanceCode(const MasterAlgorithm *const pAlgorithm) const
+{
+    const MCParticleList *pMCParticleList = nullptr;
+    PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetCurrentList(*pAlgorithm, pMCParticleList));
+
+    MCParticleVector trueNeutrinos;
+    LArMCParticleHelper::GetTrueNeutrinos(pMCParticleList, trueNeutrinos);
+
+    if (trueNeutrinos.size() != 1)
+    {
+        std::cout << "NeutrinoIdTool::GetNuanceCode - Error: number of true neutrinos in event must be exactly one" << std::endl;    
```

Comment:
I did find a few events with multiple true neutrinos, with valid nuance codes.

First of all, is this OK with MCNeutrinos that don't have nuance codes - presumably LArMCParticleHelper doesn't pass these as true neutrinos?

Secondly, is everything OK if the exception is raised here? What happens, is the event just skipped from training?

Obvious question - no MC information is ever even collected for regular use case i.e. classification, right? :)

Reply:
1. Yes the LArMCParticleHelper insists that the neutrino has a nuance code
https://github.com/PandoraPFA/LArContent/blob/5a4156a0e7a0e7ddaad3d378a0a0a3f42204d5f7/larpandoracontent/LArHelpers/LArMCParticleHelper.cc#L83
2. You are right, no MC information is ever collected in the regular use case. If an exception is raised then the event will be skipped altogether from training. The idea is if you run this on a cosmic only sample in training mode it wouldn't produce any training samples at all.